File: DSAlgorithms/Leetcode/c++/binary_search/1011-ship_packages.cpp
class Solution {
public:
    bool can_ship(int mid, int days, std::vector<int>& weights)
    {
        int ships = 1;
        int current_cap = mid;
        for(int i = 0; i < weights.size(); i++)
        {
            if((current_cap - weights[i]) < 0)
            {
                ships++;
                current_cap = mid;
            }
            current_cap -= weights[i];
        }
        return ships <= days;
    }
    int shipWithinDays(vector<int>& weights, int days) {
       int left = 0;
       int right = 0; 
       for(int i = 0; i < weights.size(); i++)
       {
           left = std::max(left, weights[i]);
           right += weights[i];
       }
       int result = right; 
       while(left <= right)
       {
           int mid = left + (right - left) / 2;
           if(can_ship(mid, days, weights))
           {
               result = std::min(result, mid);
               right = mid - 1;
           }
           else
            left = mid + 1;
       }
       return result;
    }
};

File: DSAlgorithms/Leetcode/ts/src/binary_search/1011-ship_packages.ts
const can_ship = (mid: number, days: number, weights: number[]): boolean =>{
    let ships = 1;
    let current_cap = mid;
    for(let i = 0; i < weights.length; i++)
    {
        if((current_cap - weights[i] < 0))
        {
            ships++;
            current_cap = mid;
        }
        current_cap -= weights[i];
    }
    return ships <= days;

}
const shipWithinDays = (weights: number[], days: number): number =>{
    let left = 0;
    let right = 0;
    for(let i = 0; i < weights.length; i++)
    {
        left = Math.max(left, weights[i]);
        right += weights[i];
    }
    let result = right;
    while(left <= right)
    {
        let mid = Math.floor(left + (right - left) / 2);
        if(can_ship(mid, days, weights))
        {
            result = Math.min(result, mid);
            right = mid - 1;
        }
        else
            left = mid + 1;
    }
    return result;
};